### PR TITLE
[CLI] add support to run app on a specific cluster

### PR DESCRIPTION
**Pull request by nicolai86**
Follow up PR for https://github.com/Lightning-AI/lightning/pull/13835.

This PR changes the run command to support a `--cluster-id` flag when `--cloud` is provided.
This way customers can run applications on their own BYOC compute clusters.

## Verification

- new unit tests
- updated existing unit tests
- working e2e test

### CODE REVIEW

**Comment by awaelchli** [comment]
@nicolai86 I suggest you rebase this branch onto the other one, then change the base branch here in github to the other one to have the reviewers only see the diff. This way, not all changes need to be reviewed again. 

image

To request a review, click "ready for review" at the bottom. Code-owners will get a notification.

File: tests/tests_app_examples/test_v0_app.py [hunk 1]
```diff
@@ -30,6 +30,42 @@ def test_v0_app_example():
...
+    with run_app_in_cloud(
+        os.path.join(_PROJECT_ROOT, "examples/app_v0"),
+        extra_args=["--cluster-id", os.environ.get("LIGHTNING_BYOC_CLUSTER_ID")],
+    ) as (
+        _,
+        view_page,
+        fetch_logs,
+    ):
+
+        def check_content(button_name, text_content):
```
[the remark this answers is not in the document]
Comment (nicolai86) [reply tests/tests_app_examples/test_v0_app.py:48]: done :+1:

File: src/lightning_app/cli/lightning_cli.py [hunk 2]
```diff
@@ -121,6 +130,9 @@ def run():
 @run.command("app")
 @click.argument("file", type=click.Path(exists=True))
 @click.option("--cloud", type=bool, default=False, is_flag=True)
+@click.option(
+    "--cluster-id", type=str, default=None, help="Run Lightning App on a specific Lightning AI BYOC compute cluster"
```
[the remark this answers is not in the document]
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:134]: @Felonious-Spellfire WDYT?

[the remark this answers is not in the document]
Comment (manskx) [reply src/lightning_app/cli/lightning_cli.py:134]: I think we can do something like:
"Run Lightning App on your own cluster. check <URL> to know more"

File: .lightningignore [hunk 3]
```diff
@@ -0,0 +1 @@
+.git
```
Comment (manskx) [inline .lightningignore:1]: Could this be removed  ?

File: src/lightning_app/CHANGELOG.md [hunk 4]
```diff
@@ -10,7 +10,6 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Add support for `Lightning App Commands` through the `configure_commands` hook on the Lightning Flow and the `ClientCommand`  ([#13602](https://github.com/Lightning-AI/lightning/pull/13602))
 - Add support for Lightning AI BYOC cluster management ([#13835](https://github.com/Lightning-AI/lightning/pull/13835))
```
Comment (manskx) [inline src/lightning_app/CHANGELOG.md:12]: nit: Maybe we can update the changelog to include this PR also

File: .lightningignore (hunk 3 above)
Comment (nicolai86) [reply .lightningignore:1]: when I run the acceptance tests locally the entire lightning repository is packaged and uploaded to Lightning AI - maybe this is due to how I run things.
I'll remove it.

File: src/lightning_app/CHANGELOG.md (hunk 4 above)
Comment (nicolai86) [reply src/lightning_app/CHANGELOG.md:12]: good point, added!

File: src/lightning_app/cli/lightning_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Felonious-Spellfire) [reply src/lightning_app/cli/lightning_cli.py:134]: @tchaton @nicolai86 

Linking to the doc would be great if we can do it.

[the remark this answers is not in the document]
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:134]: Can we follow up on just the doc string in a separate PR? I don't know how to get a proper URL without https://github.com/Lightning-AI/lightning/pull/13976 being merged & released.

**Review by manskx — APPROVED** [review]

File: src/lightning_app/runners/cloud.py [hunk 5]
```diff
@@ -157,14 +160,21 @@ def dispatch(
                 enable_app_server=app_spec.enable_app_server,
                 flow_servers=app_spec.flow_servers,
                 image_spec=app_spec.image_spec,
+                cluster_id=cluster_id,
```
Comment (hhsecond) [inline src/lightning_app/runners/cloud.py:163]: I am not sure coupling app releases with a cluster is the right thing to do here. Apps releases are like new app versions that's available in the app store. The cluster coupling comes up when somebody tries to run this app version and hence should be an attribute of app instance (few lines below this)

File: src/lightning_app/cli/lightning_cli.py [hunk 6]
```diff
@@ -142,7 +155,7 @@ def run_app(
     app_args: List[str],
 ):
     """Run an app from a file."""
-    _run_app(file, cloud, without_server, no_cache, name, blocking, open_ui, env)
+    _run_app(file, cloud, cluster_id, without_server, no_cache, name, blocking, open_ui, env)
```
Comment (hhsecond) [inline src/lightning_app/cli/lightning_cli.py:158]: can we raise if `cluster-id` is provided without `--cloud`

File: src/lightning_app/runners/cloud.py [hunk 7]
```diff
@@ -108,6 +110,7 @@ def dispatch(
                 random_name = "".join(random.choice(string.ascii_lowercase) for _ in range(5))
                 spec = V1LightningworkSpec(
                     build_spec=build_spec,
+                    cluster_id=cluster_id,
```
Comment (hhsecond) [inline src/lightning_app/runners/cloud.py:113]: `work.cloud_compute` already has a `clusters` attribute. We should check if there is a conflict between the values there and the `--cluster-id` provided and then either raise or continue with the priority one but that's probably a decision we need from Product

**Review by hhsecond — CHANGES_REQUESTED** [review]
Thanks for the PR @nicolai86. Overall, the flow looks good

I have made few comments about some of the early design choices we made which conflicts with the design of this PR. We don't necessarily have to go with the original design but we need to come to a conscious consensus on how we should go about it

File: src/lightning_app/runners/cloud.py (hunk 5 above)
Comment (hhsecond) [reply src/lightning_app/runners/cloud.py:163]: I am not even sure why we have `cluster-id` in the release spec at all

File: src/lightning_app/cli/lightning_cli.py (hunk 6 above)
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:161]: done, and added unit tests for it.

File: src/lightning_app/runners/cloud.py (hunk 7 above)
Comment (nicolai86) [reply src/lightning_app/runners/cloud.py:113]: discussed internally - for now let's not rely on `work.cloud_compute.clusters` as we don't support the suggested functionality of distributing work between different clusters today

File: src/lightning_app/runners/cloud.py (hunk 5 above)
Comment (nicolai86) [reply src/lightning_app/runners/cloud.py:163]: From manual verification, the cluster didn't start on the selected cluster without passing this parameter to the release.

I think we need to circle back on our service design and omit this in the API if we don't need it. For now it's required to support this sadly.

File: src/lightning_app/runners/cloud.py [hunk 8]
```diff
@@ -157,14 +160,21 @@ def dispatch(
...
             )
+            if cluster_id is not None:
+                self._ensure_cluster_project_binding(project.project_id, cluster_id)
+
             lightning_app_release = self.backend.client.lightningapp_v2_service_create_lightningapp_release(
                 project.project_id, lightning_app.id, release_body
             )
 
+            if cluster_id is not None:
+                print(f"running app on {lightning_app_release.cluster_id}")
```
Comment (tchaton) [inline src/lightning_app/runners/cloud.py:176]: Let's use logger and provide a better message to the users.

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_app/cli/lightning_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_app/cli/lightning_cli.py:137]: How does the cluster id relate to the cluster name given in the cloud compute? 

Also, @nicolai86 can you give us an example of such an ID? Is it human readable or just a long number?

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/runners/cloud.py (hunk 8 above)
Comment (nicolai86) [reply src/lightning_app/runners/cloud.py:176]: good point, changed to logger.

File: src/lightning_app/cli/lightning_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli.py:137]: today the cluster id is the human readable name provided by the customer when the cluster is created.

so it's anything within `[a-zA-Z][a-zA-Z-]+`, in the control of the cluster creator.

ie for Lightning we use `litng-ai-03`

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_app/cli/lightning_cli.py:137]: @tchaton That means we will remove the cluster argument from the CloudCompute?

**Pull request merged**